docs/architecture/adr-041-in-place-store-migrations.md

```diff
@@ -0,0 +1,174 @@
+type MigrationKeeper interface {
+    // Migrate1 migrates the store from version 1 to 2.
+    Migrate1(ctx sdk.Context) error
+
+    // ...Add more MigrateN scripts here.
+}
+```
+
+Since migration scripts manipulate legacy code, they should live inside the `legacy/` folder of each module, and be called by the keeper's implementation of `MigrationKeeper`.
+
+```go
+func (keeper BankKeeper) Migrate1(ctx sdk.Context) error {
+    return v042bank.MigrateStore(ctx, keeper.storeKey) // v042bank is package `x/bank/legacy/v042`.
+}
+```
+
+Each module's migration scripts are specific to the module's store evolutions, and are not described in this ADR. An example of x/bank store key migrations following the introduction of ADR-028 length-prefixed addresses can be seen [here](https://github.com/cosmos/cosmos-sdk/blob/ef8dabcf0f2ecaf26db1c6c6d5922e9399458bb3/x/bank/legacy/v042/store.go#L15).
+
+### Tracking Module Versions in `x/upgrade`
+
+We introduce a new prefix store in `x/upgrade`'s store. This store will track each module's current version, it can be modelized as a `map[string]uint64` of module name to module ConsensusVersion, and will be used when running the migrations (see next section for details). The key prefix used is `0x1`, and the key/value format is:
+
+```
+0x1 | {module_name_bytes} => LittleEndian(module_consensus_version)
+```
+
+We add a new parameter `ModuleManager` `x/upgrade`'s `NewKeeper` constructor, where ModuleManager is:
+
+```go
+type ModuleManager interface {
+    GetConsensusVersions() MigrationMap
+}
+
+// Map of module name => ConsensusVersion.
+type MigrationMap map[string]uint64
+```
+
+When `InitChain` is run, the initial `ModuleManager.GetConsensusVersions()` value will be recorded to state. The UpgradeHandler signature needs to be updated to take a `MigrationMap`, as well as return an error:
+
+```diff
+- type UpgradeHandler func(ctx sdk.Context, plan Plan)
++ type UpgradeHandler func(ctx sdk.Context, plan Plan, migrationMap MigrationMap) error
+```
+
+Applying an upgrade will fetch the `MigrationMap` from the `x/upgrade` store and pass it into the handler.
+
+```diff
+func (k UpgradeKeeper) ApplyUpgrade(ctx sdk.Context, plan types.Plan) {
+    // --snip--
+-   handler(ctx, plan)
++   handler(ctx, plan, k.GetConsensusVersions()) // k.GetConsensusVersions() fetches the MigrationMap stored in state.
+}
+```
+
+### Running Migrations
+
+Once all the migration handlers are registered inside the configurator (which happens at startup), running migrations can happen by calling the `RunMigrations` method on `module.Manager`. This function will loop through all modules, and for each module:
+
+- Fetch the old ConsensusVersion of the module from its `migrationMap` argument (let's call it `M`).
+- Fetch the new ConsensusVersion of the module from the `ConsensusVersion()` method on `AppModule` (call it `N`).
+- If `N>M`, run all registered migrations for the module sequentially `M -> M+1 -> M+2...` until `N`.
```

Comment:
Let's explain that if there is no old ConsensusVersion, no migration is run but the version is set in the new migration map. i.e. if we add module `foo` to our chain at version 5, no migrations will be run but 5 will be saved as the current version after migrations are run.

Reply:
Done. The `upgradeKeeper.SetCurrentConsensusVersions()` method to store the current version is run right after the upgrade handler.

btw, I removed saving ConsensusVersions during InitChain, because we don't need it.